src/main/java/com/slack/kaldb/metadata/core/CacheableMetadataStore.java

```diff
@@ -0,0 +1,101 @@
+package com.slack.kaldb.metadata.core;
+
+import com.slack.kaldb.metadata.zookeeper.CachedMetadataStore;
+import com.slack.kaldb.metadata.zookeeper.CachedMetadataStoreListener;
+import com.slack.kaldb.metadata.zookeeper.MetadataStore;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.state.ConnectionState;
+import org.slf4j.Logger;
+
+/**
+ * This store adds a cache on top of the KaldbMetadataStore. Optionally, this class can disable
+ * caching by passing in a feature flag.
+ *
+ * <p>Instantiating an instance of this class creates the cache, adds a listener and starts the
+ * cache. The listener doesn't notify the watchers on cache initialization but fires on subsequent
+ * cache events. To close the cache, call the close method.
+ */
+public abstract class CacheableMetadataStore<T extends KaldbMetadata>
+    extends KaldbMetadataStore<T> {
+
+  private final Optional<CachedMetadataStore<T>> cache;
+
+  private final List<KaldbMetadataStoreChangeListener> watchers;
+
+  public CacheableMetadataStore(
+      boolean shouldCache,
+      String snapshotStoreFolder,
+      MetadataStore metadataStore,
+      MetadataSerializer<T> metadataSerializer,
+      Logger logger)
```

Comment:
Why are we passing in the logger instead of initializing it in this class? This makes invoking logging inconsistent (ie, some cases you need to do `LOG.error()` vs `logger.error()`.